### PR TITLE
feat(imgproc): introduce CubeCL GPU backend and memory mapping PoC

### PR DESCRIPTION
**Pull request by AnilKumarSingh9856**
## 📝 Description

**⚠️ Issue Link Required**: This PR must be linked to an approved and assigned issue. See [Contributing Guide](CONTRIBUTING.md#pull-request) for details.

**Fixes/Relates to:** N/A - Opened as a Draft Proof of Concept for GSoC 2026 architectural discussion prior to formal issue creation.

**Important**:
- Ensure you are assigned to the linked issue before submitting this PR
- This PR should strictly implement what the linked issue describes
- Do not include changes beyond the scope of the linked issue

This is a **Draft PR** establishing the Proof of Concept for the GSoC 2026 `kornia-imgproc` GPU acceleration project.

Rather than modifying the core CPU pipeline, this introduces the CubeCL framework as a strictly optional dependency behind the `gpu` feature flag, ensuring zero bloat for default CPU users. To validate the host-to-device memory mapping architecture, I have implemented a WGPU-accelerated `rgb_to_grayscale` kernel.

@edgarriba - I am opening this as a Draft to request your architectural review on the memory extraction approach (`gray_from_rgb_cubecl`) and Cargo feature gating before I scale this backend to `resize` and the remaining color conversions.

---

## 🛠️ Changes Made
- [x] Added `cubecl` (v0.9.0) and `bytemuck` as optional dependencies isolated strictly under `[features] gpu`.
- [x] Implemented `gray_from_rgb_cubecl` to safely intercept Kornia's 1D flat memory layout via `Image::as_slice()`, casting to byte vectors for the PCIe bus transfer without disrupting the existing CPU `Image` abstractions.
- [x] Configured dynamic hardware scaling: Calculates `CubeCount` based on image size and hardcodes `CubeDim` to 256 threads (8 warps) to maximize Streaming Multiprocessor occupancy and prevent register exhaustion.

---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** Added `test_gray_from_rgb_cubecl` which mathematically verifies the GPU 32-bit float output against the BT.601 standard down to a `1e-6` tolerance.
- [x] **Manual Verification:** Executed `cargo clippy --workspace --all-targets --features gpu -- -D warnings` and the standard `kornia-imgproc` test suite to ensure the CPU pipeline remains untouched and regression-free.
- [x] **Performance/Edge Cases:** Ran Criterion benchmarks (`benches/bench_color.rs`) on a massive 4K resolution image (`3840x2160`) to explicitly profile the WGPU initialization and PCIe transfer latency against Kornia's optimized `ndarray::Zip` CPU baseline.

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [ ] I am assigned to the linked issue (required before PR submission)
- [ ] The linked issue has been approved by a maintainer
- [ ] This PR strictly implements what the linked issue describes (no scope creep)
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

## 💭 Additional Context
**Benchmark Reality & Strategy:**
A local Criterion benchmark on a 4K image (`3840x2160`) reveals the expected hardware bottleneck for isolated operations:
* CPU (`ndarray::Zip`): ~7.0 ms
* GPU (`cubecl::wgpu`): ~85.0 ms

The standalone WGPU initialization and PCIe transfer overhead currently outweighs the parallel compute speed for a single call. However, the purpose of this PR is not to replace the CPU for single operations, but to establish the memory bridge. Once this `.as_slice()` pipeline is approved, subsequent PRs will chain operations (e.g., `resize -> grayscale -> sobel`) entirely on the device, eliminating repeated PCIe transfers and bypassing the CPU entirely.

### CODE REVIEW

File: crates/kornia-imgproc/Cargo.toml [hunk 1]
```diff
@@ -85,6 +85,9 @@ rayon = { workspace = true }
 thiserror = { workspace = true }
 wide = "1.1.1"
 
+cubecl = { version = "0.9.0", optional = true, features = ["wgpu"]}
```
[the remark this answers is not in the document]
Comment (AnilKumarSingh9856) [reply crates/kornia-imgproc/Cargo.toml]: **Accepted.** Disabled default features to strictly isolate the build to `wgpu`, preventing unnecessary CUDA/HIP driver bloat when the feature is requested.

File: crates/kornia-imgproc/src/color/gray.rs [hunk 2]
```diff
@@ -203,6 +203,91 @@ where
...
+        let g = input[start_idx + 1];
+        let b = input[start_idx + 2];
+
+        output[pixel_idx as usize] = (r * 0.299) + (g * 0.587) + (b * 0.114);
+    }
+}
+
+/// Convert an RGB image to grayscale using CubeCL GPU acceleration.
+#[cfg(feature = "gpu")]
+pub fn gray_from_rgb_cubecl<A1: ImageAllocator, A2: ImageAllocator>(
```
[the remark this answers is not in the document]
Comment (AnilKumarSingh9856) [reply crates/kornia-imgproc/src/color/gray.rs:244]: **Accepted.** Expanded the rustdoc with `# Arguments`, `# Returns`, `# Errors`, and specific size mismatch boundary conditions for the CubeCL backend.

File: crates/kornia-imgproc/src/color/gray.rs [hunk 3]
```diff
@@ -203,6 +203,91 @@ where
...
+
+    let output = client.empty(std::mem::size_of_val(dst_slice));
+
+    // 5. Calculate Hardware Grid/Block Sizes
+    // (Hint: Use block_size = 256, calculate grid_size based on out_len)
+    let block_size = 256;
+    let grid_size = (out_len as u32).div_ceil(block_size);
+
+    // 6. Launch the Kernel (unsafe block)
+    unsafe {
```
[the remark this answers is not in the document]
Comment (AnilKumarSingh9856) [reply crates/kornia-imgproc/src/color/gray.rs]: **Accepted.** Added an explicit `// SAFETY:` contract detailing the memory lifetime invariants and boundary checks to comply with `AGENTS.md`. Furthermore, I am now propagating the kernel compilation/launch `Result` up to the public API as an `ImageError::Other` instead of silently ignoring hardware faults.

File: crates/kornia-imgproc/src/color/gray.rs [hunk 4]
```diff
@@ -203,6 +203,91 @@ where
...
+            dst.cols(),
+            dst.rows(),
+        ));
+    }
+
+    // 2. Initialize the WGPU Runtime and Client
+    type Runtime = cubecl::wgpu::WgpuRuntime;
+    let device = Default::default();
+    let client = Runtime::client(&device);
+
```
[the remark this answers is not in the document]
Comment (AnilKumarSingh9856) [reply crates/kornia-imgproc/src/color/gray.rs]: **Acknowledged, but intentionally keeping as-is for this PoC.** While caching the client in a global `OnceLock` solves the immediate micro-benchmark overhead, hiding hardware state in static globals is an architectural anti-pattern for a library. This PR strictly scopes the runtime locally to isolate and verify the `.as_slice()` memory mapping. The final GSoC architecture will introduce a dedicated `Context` or `Pipeline` struct to persist the WGPU client across chained operations, rather than relying on hidden static state.

File: crates/kornia-imgproc/src/color/gray.rs [hunk 5]
```diff
@@ -203,6 +203,91 @@ where
...
+    // 3. Extract the flat CPU arrays from Kornia's Image wrapper
+    let src_slice = src.as_slice();
+    let dst_slice = dst.as_slice_mut();
+    let out_len = dst_slice.len();
+
+    // 4. Handle Host-to-Device Memory Transfer
+    // (Hint: Convert src_slice to Bytes, call client.create and client.empty)
+    let input_bytes = Bytes::from_bytes_vec(cast_slice(src_slice).to_vec());
```
[the remark this answers is not in the document]
Comment (AnilKumarSingh9856) [reply crates/kornia-imgproc/src/color/gray.rs]: **Accepted.** Removed tutorial-style scaffolding notes and replaced them with production-grade architectural comments.

File: crates/kornia-imgproc/src/color/gray.rs [hunk 6]
```diff
@@ -360,4 +445,49 @@ mod tests {
...
+            ],
+            CpuAllocator
+        )?;
+
+        // 2. Allocate the empty destination image
+        let mut gray = Image::<f32, 1, _>::from_size_val(image.size(), 0.0, CpuAllocator)?;
+
+        // 3. Execute your new CubeCL GPU function
+        super::gray_from_rgb_cubecl(&image, &mut gray)?;
+
```
[the remark this answers is not in the document]
Comment (AnilKumarSingh9856) [reply crates/kornia-imgproc/src/color/gray.rs:485]: **Accepted.** Added `#[ignore]` to prevent headless CI runner crashes, restricting this hardware-dependent test to local verification.

File: crates/kornia-imgproc/src/color/gray.rs [hunk 7]
```diff
@@ -203,6 +203,91 @@ where
...
+            CubeDim {
+                x: block_size,
+                y: 1,
+                z: 1,
+            },
+            ArrayArg::from_raw_parts::<f32>(&input, src_slice.len(), 1),
+            ArrayArg::from_raw_parts::<f32>(&output, out_len, 1),
+            cubecl::frontend::ScalarArg::new(out_len as u32),
+        );
+    }
```
[the remark this answers is not in the document]
Comment (AnilKumarSingh9856) [reply crates/kornia-imgproc/src/color/gray.rs]: **Accepted.** Added an explicit `// SAFETY:` contract detailing the memory lifetime invariants and boundary checks to comply with `AGENTS.md`. Furthermore, the code now explicitly captures the kernel compilation/launch `Result` and propagates it up to the public API as an `ImageError::Other`, rather than silently ignoring potential hardware faults.

**Comment by sidd-27** [comment]
make this on yout own fork and share the link, dont open prs please

**Comment by AnilKumarSingh9856** [comment]
> make this on yout own fork and share the link, dont open prs please

Understood. My apologies for cluttering the upstream PR queue. I will keep the PoC strictly on my fork and share the branch link in the Discord/Discussions for architectural review. Thanks for the heads-up!
